tools/python/mwm_downloader.py

```diff
@@ -0,0 +1,135 @@
+ROOT = os.path.dirname(os.path.realpath(__file__))
+DEFAULT_DOWNLOAD_DIRECTORY = os.path.join(os.path.dirname(os.path.dirname(ROOT)), 'data')
+MAP_REVISION_REGEXP = re.compile(r'(\d{6})')
+MWM_NAME_REGEXP = re.compile(r'"(\S+).mwm"')
+
+logger = logging.getLogger(__name__)
+
+
+def parse_html_by_regexp(url, regexp):
+    parse_result = []
+    with urllib.request.urlopen(url) as f:
+        lines = f.read().decode('utf-8').split('\n')
+        for line in lines:
+            match = regexp.search(line)
+            if match:
+                parse_result.append(match.group(1))
+    return parse_result
+
+
+def get_possible_mwm_revisions():
+    return [int(revision) for revision in parse_html_by_regexp(DIRECT_MAP_URL, MAP_REVISION_REGEXP)]
+
+
+def get_possible_mwm_names(revision_num):
+    revision_url = '{base}{revision}'.format(base=DIRECT_MAP_URL, revision=revision_num)
+    return [name.replace('%20', ' ') for name in parse_html_by_regexp(revision_url, MWM_NAME_REGEXP)]
+
+
+def download_file(url, filename, attempts=3):
+    try:
+        urllib.request.urlretrieve(url, filename=filename)
+    except:
+        logger.error('Map with name {name} not loaded'.format(name=os.path.basename(filename)), exc_info=True)
+        if attempts:
+            download_file(url, filename, attempts-1)
+        return False
+    return True
+
+
+def progress_bar(total, progress):
+    """
+    Displays or updates a console progress bar.
+
+    Original source: https://stackoverflow.com/a/15860757/1391441
+    """
+    initial_progress = progress
+    bar_length = 20
+    progress = float(progress) / float(total)
+    block = int(round(bar_length * progress))
+    text = '\r[{}] {:.0f}% {}/{}'.format(
+        '#' * block + '-' * (bar_length - block), round(progress * 100, 0), initial_progress, total)
+    sys.stdout.write(text)
+    if progress == 1.:
+        sys.stdout.write('\n')
+    sys.stdout.flush()
+
+
+def download_mwm_list(revision, mwm_list, threads, folder, silent=True):
+    possible_revisions = get_possible_mwm_revisions()
+
+    if revision is None:
+        # Get latest revision by default
+        revision = max(possible_revisions)
+    elif revision not in possible_revisions:
+        logger.error('{revision} is not found in {url}'.format(revision=revision, url=DIRECT_MAP_URL))
+        return 1
+
+    possible_mwm_names = get_possible_mwm_names(revision)
+    if mwm_list is not None:
+        for mwm_name in mwm_list:
+            if mwm_name not in possible_mwm_names:
+                logger.error('{mwm_name} map is not found in {url}'.format(mwm_name=mwm_name, url=DIRECT_MAP_URL))
+                return 1
+    else:
+        mwm_list = possible_mwm_names
+
+    folder = folder if folder is not None else '{base}/{revision}'.format(base=DEFAULT_DOWNLOAD_DIRECTORY,
+                                                                          revision=revision)
+
+    if not os.path.exists(folder):
```

Comment:
It is not multiproc safe and not very pythonic. Please use https://docs.python.org/3/glossary.html#term-eafp principle

Reply:
Thanks! Fixed.